### ISSUE TITLE
error in getModelFromHomology

### ISSUE THREAD

**Issue by lucaz88** [SysBioChalmers/RAVEN#120]
Hi,
I am trying to build a draft model for an organism of interest with the function getModelFromHomology.
The code I am running so far is:

```
blastStructure = getBlast('new_genome', 'new_genome.faa', {'old_genome'}, {'old_genome.faa'});
templateModelList{1} = old_model;
SMdraftModel = getModelFromHomology(templateModelList, blastStructure, 'new_genome');
```
But I am getting the following error:

> Error using removeGenes>canRxnCarryFlux (line 104)
> Error: Invalid expression. When calling a function or indexing a variable, use parentheses. Otherwise, check for mismatched
> delimiters.
> 
> Error in removeGenes (line 62)
>                     canCarryFlux(j) = canRxnCarryFlux(reducedModel,grRule,genes{i});
> 
> Error in getModelFromHomology (line 312)
>     models{useOrderIndexes(i)}=removeGenes(models{useOrderIndexes(i)},~a,true,true,false);

I started to look into the script lines to see what could be the issue but without success. Can you help me to figure out a solution?

I am using RAVEN 2.0 and running it on Ubuntu 18.04.

Cheers,
Luca

**Comment by BenjaSanchez** [comment]
@lucaz88 sounds like some grRule that the function `canRxnCarryFlux` is receiving might be non-compliant, as the line `canIt = eval(geneRule);` is failing. Have you tried as a first step in your pipeline running your `old_model` through `standardizeGrRules.m` to check if all rules are properly defined? The function should detect which rules are non-compliant and suggest you how to change them:

```matlab
[grRules,rxnGeneMat,indexes2check] = standardizeGrRules(old_model);
```

If after changing those rules (if any) you still get the same error then please remove the final `;` in line #103 of `removeGenes` and paste here which rule is creating the error. Let us know how it goes! :)

**Comment by lucaz88** [comment]
Thanks a lot for the quick reply!!
So I ran the standardizeGrRules and it just throws me a warning for this rule:

> grRule #4239: (311550) and (309435 or 311846) and (300949 or 307814) and (306012)

But it doesn't seem to be the problem. If I completely remove that rule and run the getModelFromHomology (without semicolon at line 103 of the removeGenes script) I got the following error:

>  ....
> geneRule =
> 
>     '(false && true)'
> 
> 
> geneRule =
> 
>     '(false && true)'
> 
> 
> geneRule =
> 
>     '(AEZ63317)'
> 
> Error using eval
> Undefined function or variable 'AEZ63317'.
> 
> Error in removeGenes>canRxnCarryFlux (line 104)
> canIt    = eval(geneRule);
> 
> Error in removeGenes (line 62)
>                     canCarryFlux(j) = canRxnCarryFlux(reducedModel,grRule,genes{i});
> 
> Error in getModelFromHomology (line 312)
>     models{useOrderIndexes(i)}=removeGenes(models{useOrderIndexes(i)},~a,true,true,false);

So I looked into old_model.genes and I found this gene called 'AEZ63317'. There are also a bunch of other genes with non-numerical names (e.g. Phatr_10581, YP_874448.1,..) and they are of course included in the old_model.grRules so I cannot remove them. Do you think that if I change the gene IDs with new numeric-names I could bypass the issue?

**Comment by BenjaSanchez** [comment]
@lucaz88 no problem with having non-numerical gene IDs, I see the problem now. Your rule comes with extra parenthesis `(AEZ63317)`, when it should be for proper parsing just `AEZ63317`. `removeGenes` requires for proper functioning that rules don't have those unnecessary parenthesis, and `standardizeGrRules` actually can take care of those. So I would then recommend:
1. Either delete grRule #4239 or make it compliant: `(311550 and 309435 and 300949 and 306012) or (311550 and 311846 and 300949 and 306012) or (311550 and 309435 and 307814 and 306012) or (311550 and 311846 and 307814 and 306012)`
2. Update the grRules in `old_model` to be compliant: 
```matlab
[grRules,rxnGeneMat,indexes2check] = standardizeGrRules(old_model);
old_model.grRules = grRules;
old_model.rxnGeneMat = rxnGeneMat;
```

And then run your pipeline. See how that goes :)

On a side note, @edkerk @simas232 @Hao-Chalmers I see that `getModelFromHomology` calls `removeGenes` with the flag `standardizeRules = false`. Is there a specific reason for that? If we would make it `true` then this problem would have been skipped and the function would have only errored at the conflicting grRule #4239

**Comment by edkerk** [comment]
@BenjaSanchez Looking in the history of `getModelFromHomology` it seems like that flag has been `false` since it's been introduced (0c280e5aa018b474e7354a6087cff82be811c95a).

When `removeGenes` is called in `getModelFromHomology`, it's in a loop with n = number of template models. Changing the `standardizeRules` flag to true here would mean that if non-standard grRules would be encountered, it wouldn't necessarily be obvious for which of the models the grRules should be adjusted.

Instead, it might be better to run `standardizeGrRules` near the beginning of `getModelFromHomology` for all template models, like:

```matlab
for i=1:length(models)
    fprintf('Standardizing grRules of template model with ID %s:\n',models{i}.id);
    [models{i}.grRules,models{i}.rxnGeneMat]=standardizeGrRules(models{i},false);
end
```

**Comment by lucaz88** [comment]
@BenjaSanchez It worked! getModelFromHomology ran without problem, although the resulting draft model has no rxns or mets.
I saw in an old reply that you gave in gitter to @eileenhannas that gene IDs in the template model and in the related fasta file has to be identical. This means that the seq headers in the fasta have to match the template_model.genes field, right?

**Comment by edkerk** [comment]
@lucaz88 indeed, it is essential that the fasta seq headers match the `template_model.genes` identifiers. If you pull the latest `devel` [version](https://github.com/SysBioChalmers/RAVEN/tree/devel), `getModelFromHomology` will check whether the gene IDs match (but you already self identified that this was not the case for you).

**Comment by edkerk** [comment]
The code from https://github.com/SysBioChalmers/RAVEN/issues/120#issuecomment-403786733 has now been implemented in `devel` PR #124.
